tpe/lib/src/Model.cc

```diff
@@ -52,12 +53,50 @@ Entity &Model::AddLink()
 }
 
 //////////////////////////////////////////////////
-Entity &Model::GetCanonicalLink()
+Entity &Model::AddModel()
 {
-  // return first link as canonical link
-  return *this->GetChildren().begin()->second;
+  std::size_t modelId = Entity::GetNextId();
+  const auto[it, success]  = this->GetChildren().insert(
+      {modelId, std::make_shared<Model>(modelId)});
+
+  it->second->SetParent(this);
+  this->ChildrenChanged();
+  return *it->second.get();
 }
 
+//////////////////////////////////////////////////
+Entity &Model::GetCanonicalLink()
+{
+  std::set<Model *> models;
```

Comment:
will a `std::set` maintain the same order of nested models?

Reply:
I don't think so. Similarly, all the children (links and nested mdoels) in the a TPE Model are stored in an `std::map` which also does not guarantee order. Will that be an issue?

Reply:
the spec calls for the first link to be the canonical link if the attribute `//model/@canonical_link` is unspecified, so if this ignores the xml order then the choice of canonical link may not be consistent

it's a bit pedantic I guess. we can make an issue for it ("TPE canonical link choice may be inconsistent") instead of spending more time on it

Reply:
ok I added some logic to mark the first link added to the model as the canonical link, and also added a todo note about handling (or preventing) removal of this link. The order in which the links are added should be consistent with the xml order so the canonical link should also be consistent now. d2a9481

Reply:
the handling of links looks good now. for a model with no links and multiple nested models, I'm not sure that it will always pick the same model since it is storing them in a `std::set`, but I'm fine with adding a `todo` comment and opening an issue instead of iterating further

Reply:
sounds good. Added note in 359ca9a4a842ab17e8096ca6a1041e00638b7205 and ticketed issue #100